### DISCUSSION TITLE
Dark Mode Based on Parent/Sibling State?

### DISCUSSION THREAD

**Discussion by owenfarrell** [tailwindlabs/tailwindcss#11928] [Help] [answered]
Is there a way to **_enable_** dark mode using existing variants in Tailwind?

When it comes to manually enabling dark mode, I get that the common approach is to enable a bit of JavaScript. I'm specifically trying to avoid the javascript and enable dark mode on specific elements via CSS. 

For example: https://play.tailwindcss.com/4b2jgN06aQ

I've tried using using the `group`/`group-even` state outright. When that didn't work I tried merging `group`/`group-even` with `dark` via a plugin based on the comment about `peer`/`group` usage [here](https://tailwindcss.com/docs/plugins#parent-and-sibling-states), but I'm not totally sure my plugin was correct.

Any help would be appreciated.

**Comment by wongjn** [comment]
For your example, you'd use `group-even:` directly with the `underline` utility: https://play.tailwindcss.com/QLaFArKpeT

As for activating class strategy dark mode via CSS in more general terms, this is impossible. You cannot add classes to elements via CSS, thus you cannot add classes to elements with Tailwind either.

**Reply by owenfarrell** [reply]
@wongjn - I should've been more clear in that my example is meant to illustrate how I would like to enable `dark`. You're absolutely right in that I can use the `group-even`/`group-odd` variants in conjunction with the `underline`.

Here's a better example of what I'm trying to achieve: https://play.tailwindcss.com/yBhCGgffX2. All of the CSS classes that I'm looking to add are on the elements. But the intended outcome isn't there.

In doing some research, I **_think_** I understand why this is happening. When `darkMode` is defined as a class, `dark` is added as a variant but is not added as a utility class itself.

https://github.com/tailwindlabs/tailwindcss/blob/27b6530599b7604c1d6208e41916394ec4837976/src/corePlugins.js#L215-L232

And that seems to be confirmed with an inline comment here:

https://github.com/tailwindlabs/tailwindcss/blob/27b6530599b7604c1d6208e41916394ec4837976/src/lib/setupContextUtils.js#L917-L927

If the `darkMode` class was a bona-fide utility class (with no particular attributes), then I think it could be used in conjunction with variants like I have in my example above. And it would also get the `darkMode` class name out of the `parasiteUtilities` list int he above workaround.

**Reply by wongjn** [reply]
I'm still not quite sure why you are trying to use `dark:` here. Again, `group-even:` should solve your needs: https://play.tailwindcss.com/V2lYlONyfg

**Reply by owenfarrell** [reply]
When everything is in one file, that's a fine approach. But in more complex scenarios, I think this would help maintain clean separations of concerns.

For me, I've got different React components - think of it as a collection of cards, where each card is comprised of multiple, other components. The card container can be group aware (with `group` and `group-even`/`group-odd`).

And the components that are used _**within**_ a card can be `dark`-aware for theming. But they aren't (and I would argue, shouldn't) need to be aware of how it's used in the context of a collection of cards - those same underlying components might be used very differently in components that have no concept of a `group`, `odd`, or `even`.

**Reply by wongjn** [reply]
You could consider having props of the React components that would allow contextual styling from a parent component.

**Reply by owenfarrell** [reply]
Would promoting the `darkMode` class from a parasite class to an (empty) utility class solve this scenario for _**all**_ apps using Tailwind (and not just React)?

(my assumption is that the answer is _**yes**_)

**Reply by wongjn** [reply]
`dark` (or the customized `darkMode` selector equivalent) does not do anything of significance on its own – it's used as a contextual selector prefix attached to other rule selectors.

**Reply by owenfarrell** [reply]
Any thoughts on a plugin configuration that might do the trick? I can't seem to figure out the magic combination of directives and selectors.

Given Tailwind's support for parent/sibling states and dark mode via classes, this feels like something that _**should**_ be achievable by combining those two capabilities.

**Reply by wongjn** [reply, marked as answer]
For how I believe you are trying to get `group-even:dark` to work, you'd need to do something that would generate a selector like:

```css
.dark .dark\:bg-black,
.group:nth-child(even) .group-even\:dark .dark\:bg-black {
```

But the thing is, you'd need for Tailwind to collate any "modified" `dark` classes like `group-even:dark` and then build an intermediary list of selector scopes to then generate and attach on to each class with the `dark:` variant class. You could also get invalid selectors like with `before:dark` (not really relevant but just something I thought about):

```css
.dark .dark\:bg-black,
.before\:dark::before .dark\:bg-black {
```

I'm not sure if building an intermediary list of selector scopes would be feasible with the current plugin architecture.

You could consider having props of the React components that would allow contextual styling from a parent component instead of trying to finagle `dark`.

**Reply by owenfarrell** [reply]
> You could consider having props of the React components that would allow contextual styling from a parent component instead of trying to finagle `dark`.

Yeah, this is basically where I landed. Reconfiguring the `dark` selector to use a data attribute (`darkMode: ["class", '[data-dark="true"]']`) keeps it in the Javascript space, but ultimately avoids the need to inject values in to a `className` attribute. In hindsight, that `className` injection was what I was really trying to avoid.
